### DISCUSSION TITLE
Debugging of the External Module, Hot Reload simulation

### DISCUSSION THREAD

**Discussion by vzdesic** [oqtane/oqtane.framework#1194] [Show and tell] [answered]
Here is **description of the minimal routine which gives us ability to simulate Hot Reload** feature when **debugging External Module**:
**1.** Start Oqtane.Server from command line: **dotnet run --project path/to/Oqtane.Server.csproj**
**2.** Open the Page with External Module in the browser
**3.** External Module -> **make simple change on the UI**, change Client project's source
**4.** External Module -> Build -> **Rebuild Solution**
**5.** DLLs and PDBs files are copied to output directory of the Oqtane Framework
**6.** Stop Kestrel by pressing **Ctrl+C** in command line
**7.** Start Oqtane.Server again from command line: **dotnet run --project path/to/Oqtane.Server.csproj**
**8.** Click - **Reload** - on the Page with External Module
**9.** **Change is visible. Attach is possible.** Find process - Oqtane.Server.exe - in case you have to attach debugger of the External Module (Runtime must be set to Server).

**NOTE:** For those using VS 2019, there is useful extension - **BuildOnSave** - which can further speed up the development process of the External Module and **eliminate step 4**: https://marketplace.visualstudio.com/items?itemName=pragmatrix.BuildOnSave. With this extension DLLs of the External Module will be automatically built and copied to Oqtane.Server output directory on every Save. You can always enable or disable this extension as needed.

**Comment by sbwalker** [comment, marked as answer]
@vzdesic this is excellent! The overarching theme of the Oqtane Philosophy is "speed" - and the developer "inner loop" is a very important aspect of this story. I am sure many developers will find the Hot Reload approach to be very helpful.

**Reply by vzdesic** [reply]
@sbwalker Here is one additional improvement related to the Hot Reload, found by one of my colleagues ( @callah79 ), which we would like to suggest to be part of the Oqtane Framework source code.

If you add this into Oqtane.Server project file:
```
<ItemGroup>
    <!-- extends watching group to include *.dll files -->
    <Watch Include="**\*.dll"/>
</ItemGroup>
```

And run project with:
`dotnet watch run`

Then Reload will happen automatically, without restarting the server, in case that new External Module's DLLs are copied to the Oqtane.Server output directory.

Thanks, Vedran

**Reply by Behnam-Emamian** [reply]
This code is going to make an infinit loop for me :-/

![image](https://user-images.githubusercontent.com/9818491/220894866-752aa46f-449f-4e6e-8558-e35b3876e882.png)

**Reply by Behnam-Emamian** [reply]
Solution:
```
<ItemGroup>
    <!-- extends watching group to include *.dll files -->
    <Watch Include="**\*.dll" Exclude="**\Microsoft.EntityFrameworkCore.*.dll;**\Oqtane.Database.*.dll;"/>
</ItemGroup>
```

**Reply by sbwalker** [reply]
@Behnam-Emamian you can submit this as a PR. Personally I do not use Hot Reload - I find it to be buggy and overly resource intensive - but I know that other developers may want to use it, so it would be useful to provide support in Oqtane.
